File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    map<TreeNode*,TreeNode*>mp;
    
    void dfs(TreeNode*root, TreeNode*par){
        if(!root)  return;
        if(!par)  par=root;
        
        mp[root]=par;
        dfs(root->left,root);
        dfs(root->right, root);
    }
    
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
//         dfs root to parent mapping
        dfs(root, NULL);
//         bfs to traverse level wise from the target node
        queue<TreeNode*>q;
        set<TreeNode*>st;
        
//         to keep count of the levels
        int cnt=0;
        
//         to store the kth level nodes(the kth dist nodes from the target)
        vector<int>res;
        
//         starting bfs from the target node to kth lvl
        q.push(target);
        
        while(!q.empty()){
           int sz=q.size();
            
            if(cnt>k)  break;
            
            while(sz--){
                TreeNode*curr=q.front();
                q.pop();
                
                if(curr->left!=NULL and !st.count(curr->left)){
                    q.push(curr->left);
                    st.insert(curr->left);
                }
                if(curr->right!=NULL and !st.count(curr->right)){
                    q.push(curr->right);
                    st.insert(curr->right);
                }
                
                if(mp.count(curr) and !st.count(mp[curr])){
                    q.push(mp[curr]);
                    st.insert(mp[curr]);
                    
                }
                st.insert(curr);
                if(cnt==k){
                    res.push_back(curr->val);
                }
                
            }
            cnt+=1;
        }
        
        return res;
    }
};

File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.mp = {}
        
    def dfs(self, root, par):
        if not root:
            return
        if not par:
            par = root

        self.mp[root] = par
        self.dfs(root.left, root)
        self.dfs(root.right, root)
        
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        self.dfs(root, None)
        q = deque()
        st = set()
        cnt = 0
        res = []

        q.append(target)

        while q:
            sz = len(q)

            if cnt > k:
                break

            for _ in range(sz):
                curr = q.popleft()

                if curr.left and curr.left not in st:
                    q.append(curr.left)
                    st.add(curr.left)
                if curr.right and curr.right not in st:
                    q.append(curr.right)
                    st.add(curr.right)

                if curr in self.mp and self.mp[curr] not in st:
                    q.append(self.mp[curr])
                    st.add(self.mp[curr])

                st.add(curr)
                if cnt == k:
                    res.append(curr.val)

            cnt += 1

        return res